### PR TITLE
Hardened malloc was not disabled for oletools when an CPU with missing flags is used

### LINKED ISSUE #2959: 2.0.25 gives Restarting (132) issue in docker
### Environment

- [x ] docker compose
- [ ] kubernetes
- [ ] docker swarm

### Version

- Version: `2.0.25`

## Description

pulled the latest build (2.0.25) and the mailu docker containers would no longer start. They would go to a status of "Restarting (132)" and continuously just restart

## Expected behaviour

Docker containers to start up and run

### PR DESCRIPTION
**Pull request by Diman0**
## What type of PR?
bug fix

## What does this PR do?
Updates oletools to also disable hardened malloc when used CPU misses flags

### Related issue(s)
- closes #2959 

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [n/a ] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

### CODE REVIEW

**Comment by Diman0** [comment]
I will set the WIP flag until I have replicated the issue locally and confirmed this PR resolves it.

**Comment by nextgens** [comment]
[x86_64.txt](https://github.com/Mailu/Mailu/files/12829585/x86_64.txt)
[arm64.txt](https://github.com/Mailu/Mailu/files/12829586/arm64.txt)

Here are the instructions in use... so we need 'avx2' for vpunpckldq, what else?

File: core/base/libs/socrate/socrate/system.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def _is_compatible_with_hardened_malloc():
         lines = f.readlines()
         for line in lines:
             # See #2764, we need vmovdqu
-            if line.startswith('flags') and ' avx ' not in line:
+            if line.startswith('flags') and ' avx ' not in line and ' avx2 ' not in line:
```
Comment (nextgens) [inline core/base/libs/socrate/socrate/system.py:69]:
```suggestion
            if line.startswith('flags') and ' avx2 ' not in line:
```
You won't find a CPU that does the later but not the former.

As currently spelt out, if either is present the test will skip. This is not what we want, we want both to be present.

Comment (nextgens) [reply core/base/libs/socrate/socrate/system.py:69]: We may want to explain in the warning line 83 that we are missing vectorized instructions

**Review by nextgens — CHANGES_REQUESTED** [review]

**Review by nextgens — DISMISSED** [review]
LGTM

**Comment by Diman0** [comment]
Please reject the PR. It does not work yet.
I think I figured out the problem.
In the Dockerfile of the base image we enable the hardened malloc:
https://github.com/Mailu/Mailu/blob/dd1de62eb0e3e0f7a6dec1de5c57c3b64bb54d9b/core/base/Dockerfile#L82

As a result, the test in system.py to enable/disable hardened malloc cannot be even reached. Start.py is already run with the hardened malloc due to LD_PRELOAD=LD_PRELOAD="/usr/lib/libhardened_malloc.so being set. So incompatible systems will have the problem that start.py immediately crashes.

I propose to 
1. set `LD_PRELOAD=` in the base image. 
2. Adapt the hardened malloc test in system.py to enable the hardened malloc via LD_PRELOAD= if requirements are met (cpu flags/features  are present)

**Comment by nextgens** [comment]
bors try

**Comment by nextgens** [comment]
bors try

**Comment by nextgens** [comment]
bors try

**Comment by nextgens** [comment]
bors try

**Comment by nextgens** [comment]
@Diman0 are we happy with this now?

We may want to double check that it works in practice: use lsof to see if all PIDs of all containers are using the library

**Comment by Dennis14e** [comment]
Just my 2 cents: Wouldn't it be nicer to have a custom environment variable like "ENABLE_HARDENED_MALLOC" instead of the user setting LD_PRELOAD themselves?
Maybe LD_PRELOAD is needed for something else too at some point or an Alpine update changes the path or something similar.

Greetings & thanks for Mailu!

**Comment by Diman0** [comment]
> Just my 2 cents: Wouldn't it be nicer to have a custom environment variable like "ENABLE_HARDENED_MALLOC" instead of the user setting LD_PRELOAD themselves? Maybe LD_PRELOAD is needed for something else too at some point or an Alpine update changes the path or something similar.
> 
> Greetings & thanks for Mailu!

The point is that hardened malloc can only be made available for the entrypoint process if LD_PRELOAD is already set when the container is started. 

@nextgens It looks good to me. I will double check that it runs as expected on the VM with an outdated cpu (qemu64) and on a VM with a modern cpu with hardened malloc enabled.

**Comment by Diman0** [comment]
@nextgens It looks good to me.
On an outdated CPU (e.g. with CPU type qemu64), I do not see the warnings in the logs and Mailu starts fine.
On a modern CPU, I do get the warning to enable hardened malloc.
When running with LD_PRELOAD (hardened malloc enabled), the output of lsof confirms this.

**Review by nextgens — APPROVED** [review]

**Comment by Diman0** [comment]
bors retry

**Comment by Diman0** [comment]
bors retry

**Comment by nextgens** [comment]
bors retry

**Comment by nextgens** [comment]
bors try

**Review by nextgens — APPROVED** [review]

**Comment by nextgens** [comment]
@mergifyio backport

**Pull request merged**